docs/source/design/ui_design.rst

```diff
@@ -0,0 +1,150 @@
+standard https://pyscf.org/code-rule.html`__.
+
+Below you can find how to run an SCF calculation for a hydrogen molecule using
+PySCF.
+
+.. code-block:: python
+
+    >>> from pyscf import gto, scf
+    >>> mol = gto.M(atom='O 0 0 0; H 0 1 0; H 0 0 1', basis='sto-3g')
+    >>> rhf = scf.RHF(mol)
+    >>> energy = rhf.kernel()
+
+Here, mol is the object (type ``pyscf.gto.mole.Mole``) representing the molecule
+and integrals with the given basis set and rhf is the object (type
+``pyscf.scf.hf.RHF``) that holds method specific information. The energy is
+computed using the ``kernel()`` function.
+
+PSI4
+====
+PSI4 is a C++/Python (70%/30%) open-source quantum chemistry package distributed
+under the LGPL3 license. PSI4 provides two different types of UI referred to as
+Psithon and PsiAPI modes. In the Psithon mode, the user writes an input file in
+a domain specific language similar to Python. In the PsiAPI mode, the user can
+write a pure Python script that interacts with PSI4 as a Python module. Since
+the latter is more relavant to our design, we show below how to run an SCF
+calculation for a hydrogen molecule using the PsiAPI mode.
+
+.. code-block:: python
+
+    >>> import psi4
+    >>> mol = psi4.geometry('H 0. 0. 0. \n H 0. 0. 1.')
+    >>> energy= psi4.energy('scf/sto-3g')    
+
+Here, mol is the molecule object, which is created using the ``psi4.geometry()``
+function and the energy is computed using the ``psi4.energy()`` function, which
+takes the method name as the required argument. Note that the user do not need
+to pass the molecule object explicitly to the subsequent energy calculation. By
+default energy is computed for the last molecule defined with the
+``psi4.geometry()`` function. Additional options can be set with the
+``psi4.set_options()`` function, which takes a Python dictionary as the required
+argument. 
+
+
+QCEngine
+========
+QCEngine is a general purpose quantum chemistry program interface. It is a
+Python library that provides a common API for quantum chemistry programs.
+QCEngine is an open-source package distributed under the Apache License 2.0.
+
+Below you can find how to run an SCF calculation for a hydrogen molecule using
+QCEngine.
+
+.. code-block:: python
+
+    >>> import qcengine as qcng
+    >>> import qcelemental as qcel
+    >>> mol = qcel.models.Molecule.from_data('H 0. 0. 0. \n H 0. 0. 1.')
+    >>> out = qcng.compute({"molecule": mol, "driver": "energy", "model": {"method": "SCF", "basis": "sto-3g"}})
+    >>> energy = out.return_result
+
+Here, mol is the molecule object (type
+``qcelemental.models.molecule.Molecule``), which is created using the
+``qcel.models.Molecule.from_data()`` function from QCElemental package. The SCF
+energy is computed using the ``qcng.compute()`` function, from QCEngine
+package. Note that, the input for the `qcng.compute` function is a Python
+dictionary with a schema defined by QCElemental.
+
+
```

Comment:
Added the list with [4c34c65](https://github.com/NWChemEx-Project/.github/pull/108/commits/4c34c651942a98b82c4234e2545b4d63de4ea976).

Reply:
What you added is a good motivation, and I think reading between the lines one can infer many of the design points, which go into the UI, but I still think it would be good to explicitly list them. What follows is from what I can recall from the discussions. This probably isn't exhaustive, nor does it necessarily reflect all the view points; so it should be expanded as appropriate.

```.rst
*********************
Design Considerations
*********************

Performance
   Performance is the motivator of NWChemEX. The UI needs to be designed in a performant manner. This primarily means the user must be able to control the parallelism through the UI.

Programmatic
   We are not interested in developing a domain specific language, nor do we want to write an interpreter/parser. 

   - We are primarily targeting Python, so the UI should be importable via a Python 
     module and be written in terms of Python functions and objects.

Light weight
   Following from performance, the UI should shuttle data into C++ as quickly as 
possible.

   - For user convenience some conversions may be necessary (*e.g.*, our `Molecule` 
     class is never going to do string parsing), but the set of such conversions 
     should be minimal, or limited to those supported by some other library (which we 
     agree to take on as a dependency).

Limited results
   In theory, there's a very large set of results a user could want. Thankfully most 
users target the same quantities. The UI should focus on those quantities.

   - The most common quantities are: energy, molecular-orbital coefficients, 
     gradients, vibrational frequencies, and geometries (including minimum energy, 
     transition states, reaction paths, etc.).

Limited inputs
  Generally speaking the inputs required for the aforementioned results tend to be the 
molecule, level of theory (method plus basis set), and method parameters. While the 
set of potential parameters is huge, again there are only a handful of commonly used 
parameters.

  - Primary targets are: algorithm variations (*e.g.* density fitting), number of 
    iterations, and convergence criteria 

User focused
   The UI targets the most common use cases, but should try to be flexible. At some 
point, harnessing the full power of the code requires forgoing the UI and dropping 
down to PluginPlay's API.

Not top level
   Modern workflows are complicated. Our UI needs to avoid assuming that it is the top 
layer in order to better integrate into existing workflows.

   - The ability to take an MPI communicator is one particularly important UI design 
     element implied by this consideration combined with the performance 
     consideration.

Not In Scope
============

Graphical user interface (GUI)
   Arguably a GUI represents the pinnacle of UX; however, we presently are focused on 
a programmatic UI. Implementing a GUI is an orthogonal task which can benefit from the 
existence of the programmatic UI.

Interfaces for driving NWChemEx
   While we want NWChemEx to be part of an ecosystem, the design on this page is 
purely focused on a UI which uses a combination of native NWChemEx and Python objects.

   - With a NWChemEx UI in place driving NWChemEx from other packages becomes easier.
   - Ideally such interfaces should be maintained on the driver's side, and not by us, 
     in order to avoid needing to weigh down NWChemEx with additional dependencies.
     Note that making a dependency optional for a user does NOT negate this as 
     NWChemEx developmers must support all optional features.
```

Reply:
Thank you, all good points to include in the document.